qbft/messages.go

```diff
@@ -214,23 +234,156 @@ func (msg *Message) GetRoot() ([]byte, error) {
-		return errors.New("message data is invalid")
-	}
-	if msg.MsgType > 5 {
-		return errors.New("message type is invalid")
+func (msg *Message) Validate(msgType types.MsgType) error {
+	//if len(msg.Identifier) == 0 {
+	//	return errors.New("message identifier is invalid")
+	//}
+	if len(msg.Input) == 0 && msgType != types.ConsensusRoundChangeMsgType {
+		return errors.New("message input data is invalid")
 	}
+	//if msg.MsgType > 5 {
+	//	return errors.New("message type is invalid")
+	//}
 	return nil
 }
 
+func (msg *Message) ToMessageHeader() (*MessageHeader, error) {
+	// TODO<olegshmuelov>: implement HashTreeRoot ssz
+	//r, err := msg.Input.HashTreeRoot()
+	//if err != nil {
+	//	return &MessageHeader{}, errors.Wrap(err, "failed to get input root")
+	//}
+	return &MessageHeader{
+		Height: msg.Height,
+		Round:  msg.Round,
+		// TODO<olegshmuelov>: implement HashTreeRoot ssz
+		InputRoot:     [32]byte{},
+		PreparedRound: msg.PreparedRound,
+	}, nil
+}
+
+// SignedMessage includes a signature over Message AND optional justification fields (not signed over)
 type SignedMessage struct {
-	Signature types.Signature
-	Signers   []types.OperatorID
-	Message   *Message // message for which this signature is for
+	Message   *Message
+	Signers   []types.OperatorID `ssz-max:"13"`
+	Signature types.Signature    `ssz-size:"96"`
+
+	RoundChangeJustifications []*SignedMessageHeader `ssz-max:"13"`
+	ProposalJustifications    []*SignedMessageHeader `ssz-max:"13"`
+}
+
+//type SignedMessage struct {
+//	Signature types.Signature
+//	Signers   []types.OperatorID
+//	Message   *Message // message for which this signature is for
+//}
+
+// MessageHeader includes just the root of the input to be decided on (to save space), used for prepare, commit and justification messages
+type MessageHeader struct {
+	Height        Height
+	Round         Round
+	InputRoot     [32]byte `ssz-size:"32"`
+	PreparedRound Round
+}
+
+// Encode returns a msg encoded bytes or error
+func (m *MessageHeader) Encode() ([]byte, error) {
+	return json.Marshal(m)
+}
+
+// Decode returns error if decoding failed
+func (m *MessageHeader) Decode(data []byte) error {
+	return json.Unmarshal(data, &m)
+}
+
+func (m *MessageHeader) GetRoot() ([]byte, error) {
+	marshaledRoot, err := m.Encode()
+	if err != nil {
+		return nil, errors.Wrap(err, "could not encode message")
+	}
+	ret := sha256.Sum256(marshaledRoot)
+	return ret[:], nil
+}
+
+// SignedMessageHeader includes a signature over MessageHeader
+type SignedMessageHeader struct {
+	Message   *MessageHeader
+	Signers   []types.OperatorID `ssz-max:"13"`
+	Signature types.Signature    `ssz-size:"96"`
+}
+
+func (signedMsg *SignedMessageHeader) GetSignature() types.Signature {
+	return signedMsg.Signature
+}
+func (signedMsg *SignedMessageHeader) GetSigners() []types.OperatorID {
+	return signedMsg.Signers
+}
+
+// Encode returns a msg encoded bytes or error
+func (signedMsg *SignedMessageHeader) Encode() ([]byte, error) {
```

Comment:
we should add a followup cleanup for all the root and encoding interfaces we use as ssz comes with them already.